### ISSUE TITLE
possible false positive corruption check in compressed loadmat()

### ISSUE THREAD

**Issue by WeatherGod** [scipy/scipy#6999]
I have a file saved using `savemat(..., do_compression=True)` that errors when loaded by `loadmat()`.
```
>>> d = loadmat('afm_ff_5min27C_20060301_v03r02.mat')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/broot/scratch/miniconda/lib/python2.7/site-packages/scipy-0.18.1-py2.7-linux-x86_64.egg/scipy/io/matlab/mio.py", line 136, in loadmat
    matfile_dict = MR.get_variables(variable_names)
  File "/home/broot/scratch/miniconda/lib/python2.7/site-packages/scipy-0.18.1-py2.7-linux-x86_64.egg/scipy/io/matlab/mio5.py", line 292, in get_variables
    res = self.read_var_array(hdr, process)
  File "/home/broot/scratch/miniconda/lib/python2.7/site-packages/scipy-0.18.1-py2.7-linux-x86_64.egg/scipy/io/matlab/mio5.py", line 252, in read_var_array
    return self._matrix_reader.array_from_header(header, process)
  File "scipy/io/matlab/mio5_utils.pyx", line 673, in scipy.io.matlab.mio5_utils.VarReader5.array_from_header (scipy/io/matlab/mio5_utils.c:7087)
  File "scipy/io/matlab/mio5_utils.pyx", line 738, in scipy.io.matlab.mio5_utils.VarReader5.array_from_header (scipy/io/matlab/mio5_utils.c:6875)
ValueError: Did not fully consume compressed contents of an miCOMPRESSED element. This can indicate that the .mat file is corrupted.
```
However, this file does load without error in MATLAB. Also, if I re-run the original program and save the file with `do_compression=False`, the uncompressed file loads just fine in both python and MATLAB. Furthermore, comparing the data loaded from both uncompressed and compressed files in MATLAB yields no differences, which is indicative that the compressed file isn't actually corrupt.

My suspicion is that somehow this is a false positive in `loadmat()`. I am working on making a concise reproducible example using npz files as a source. What is the preferred method of posting data files?

**Comment by pv** [comment]
You can drag-and-drop upload files in the comment box.

**Comment by WeatherGod** [comment]
[testdata.zip](https://github.com/scipy/scipy/files/748630/testdata.zip)
Ok, if you unzip this, you get ff19d.npy. The following will reproduce the problem:
```
import numpy as np
from scipy.io import savemat, loadmat
ff19d = np.load('ff19d.npy')
savemat('ff19d_compressed.mat', dict(x=ff19d), do_compression=True)
loadmat('ff19d_compressed.mat')
```

**Comment by WeatherGod** [comment]
btw, I have some cycles to try and get this fixed, but I have never tried debugging compression code before. Any guidance/tips?

**Comment by pv** [comment]
Not really tip, but perhaps the compression itself is fine (it's zlib), 
and the issue is with bookkeeping (eg putting wrong size in headers etc).

**Comment by WeatherGod** [comment]
Did a bit more digging. It does seem that the reading aspect of all of this isn't at fault, it seems? I broke up the reading into individual steps to see if I can find deviations. I also turned off the validation testing to see if the data read are identical... they are.
```
from __future__ import print_function
import numpy as np
from scipy.io import matlab

good = 'ff19d_uncompressed.mat'
bad = 'ff19d_compressed.mat'

goodfh = matlab.mio5.MatFile5Reader(open(good, 'rb'))
badfh = matlab.mio5.MatFile5Reader(open(bad, 'rb'))

goodfh.initialize_read()
badfh.initialize_read()

good_mdict = goodfh.read_file_header()
bad_mdict = badfh.read_file_header()

mdtypeg, byte_countg = goodfh._file_reader.read_full_tag()
mdtypeb, byte_countb = badfh._file_reader.read_full_tag()

next_posg = goodfh.mat_stream.tell() + byte_countg
next_posb = badfh.mat_stream.tell() + byte_countb

streamb = matlab.mio5.ZlibInputStream(badfh.mat_stream, byte_countb)
badfh._matrix_reader.set_stream(streamb)
mdtypeb, byte_countb = badfh._matrix_reader.read_full_tag()

goodfh._matrix_reader.set_stream(goodfh.mat_stream)

headerg = goodfh._matrix_reader.read_header(False)
headerb = badfh._matrix_reader.read_header(False)  # Set to True to turn validation back on

print([x.is_global for x in headerg, headerb])
print([x.is_logical for x in headerg, headerb])
print([x.dims for x in headerg, headerb])
print([x.mclass for x in headerg, headerb])
print([x.name for x in headerg, headerb])
print([x.nzmax for x in headerg, headerb])

resg = goodfh.read_var_array(headerg, True)
resb = badfh.read_var_array(headerb, True)

goodfh.mat_stream.seek(next_posg)
badfh.mat_stream.seek(next_posb)

np.testing.assert_allclose(resg, resb, 0, equal_nan=True)
```
So, I am guessing that this is probably a write issue, possibly of the header somehow. I did see that the read automatically snapped to 64-bit boundaries, so maybe it is somehow correcting for a bad header info.

**Comment by WeatherGod** [comment]
I modified the above code a bit to have it do a `tell()` at the end of reading the file. The uncompressed file ended up at byte 2133968, and the compressed file ended up at 655496. That is correct for the uncompressed file, but incorrect for the compressed file (2 bytes short).

**Comment by cdowen** [comment]
Another sample file that will produce the problem. Tested on scipy 1.1.0, python 3.6.6 anaconda
[OF_012.mat.zip](https://github.com/scipy/scipy/files/2593540/OF_012.mat.zip)

**Comment by cdowen** [comment]
Another sample file that will produce the problem. Tested on scipy 1.1.0, python 3.6.6 anaconda
[OF_012.mat.zip](https://github.com/scipy/scipy/files/2593548/OF_012.mat.zip)

**Comment by walshb** [comment]
The commit above fixes this bug (pull request here: https://github.com/scipy/scipy/pull/10158).

ZlibInputStream stops reading the compressed stream when it has output all the uncompressed data, and throws the "Did not fully consume" error if there are bytes left in the compressed stream. But there can be a checksum at the end of the compressed stream. ZlibInputStream reads data in blocks, so if the final block only contains just the checksum, it won't be consumed and the error will be thrown.

I made sure ZlibInputStream reads the checksum (if any). I also made sure that the checksum is checked, even if check_stream_limit is False. This makes the logic the same whether the checksum is in the same final block as the data or is in the final block alone.
